### DISCUSSION TITLE
How would linear types work with an open ended system?

### DISCUSSION THREAD

**Discussion by hrj** [austral/austral#368] [Q&A]
Imagine this function (in pseudo-code):

```
file = Open a file
while (!done) {
  ch = read the next character from STDIN
  if (ch == 'x') {
    closeFile(file)
  } else {
    readFile(file)
  }
  if (some other exit condition) {
    done = true
  }
}
if (file not closed) {
  closeFile(file)
}
```

How would you write this with linear types?

**Comment by anandology** [comment]
I'm yet to understand how that works with loops, but I can easily think of a recursive solution to address that.

```
// pseudo code - tried to be as close to Austral as possible

function readLines(path, lines) is 
    let f: File! = openFile(path);
   let (line, f1) = readLine(f)
   if line == "STOP\n" then
      closeFile(f1)
  else
      lines.append(line) 
     readLines(f1, lines)
  end if;
  return nil
end;
```

## Explanation

As per my understanding, one distinct way linear types are different is that they are single use only and that is the only thing that the compiler enforces. 

So if you call a function like `read` or `readLine`, that consumes the file. Since it is an object of linear type, it can't be used again. So the `readLine` will have have to return a new handle to the file, that can again be read or closed. In the above example, I've assigned it variable `f1` and that is either close or passed to `readLines` recursively.  If you notice, both `f` and `f1` are used exactly once.

Please look at the [Linear Types section in the specifications][1] for more explanation.

[1]: https://austral-lang.org/spec/spec.html#rationale-linear-types

**Reply by asdofindia** [reply]
```
readLines(f1, lines)
```

This f1 is never consumed, right?

**Reply by anandology** [reply]
As per my understanding, passing a value of linear type as an argument to a function is considered as a used. And it can't be used again after that. 

See [Introduction to Linear Types][1] for more explanation.

https://austral-lang.org/linear-types

**Comment by eudoxia0** [comment]
Consider the following API. I'm eliding error handling for simplicity, but imagine a smattering of `Option` types:

```austral
type File: Linear;

function openFile(path: String): File;

generic [R: Region]
function isEmpty(ref: &[File, R]): Bool; -- `True` when the stream is EOF

generic [R: Region]
function readChar(ref: &![File, R]): Nat8;

function closeFile(file: File): Unit; -- consumes the file
```

A simple example: read a file char by char until EOF:

```austral
let file: File := openFile(path);
while not isEmpty(&file) do
    let char: Nat8 := readChar(&!file);
    -- do something with `char`
end while;
closeFile(file);
```

A more complex example, with multiple conditions to stop reading the file:

```austral
let file: File := openFile(path);
let condition: Bool := True;
while condition do
    let char: Nat8 := readChar(&!file);
    -- quit when we reach EOF
    if isEmpty(&file) do
        condition := False;
    end if;
    -- quit under some other condition
    if somePredicate(char) do
        condition := False
    end if;
    -- do something with `char`
end while;
closeFile(file);
```

**Reply by eudoxia0** [reply]
There is a (recursive) way to do this with pure linear types, but it's annoying, so we use borrowing instead.

**Reply by hrj** [reply]
Thanks for the reply. My original question wasn't very well phrased. I wanted to create an example of non-deterministic flow, that a compiler will not be able to predict, and hence would not be able to provide guarantees about. For example, it wouldn't be able to provide the guarantee related to use-after-free error, if it can't predict when the free will happen at compile time.

Here is a better illustrated example:

Implement an application that reads commands from STDIN. The possible commands are:
    O: open file named "input.txt"
    R : read from file, if it is open
    C: close file
    P: print the last read character to STDIN
    Q: Quit
    Any other character: NOP

The user could enter the commands in any order, such as `ORRRCRPQ`

How would you implement this with a filesystem API with linear types?

---

Aside, I later realized that for specific scenarios like this the language could provide a different API which is not based on Linear types. Or alternatively, it could have an `unsafe` block in which linear types are not enforced.

**Reply by eudoxia0** [reply]
Oh, I see. Do you mean something like:

```
let foo: Foo := make(); -- `Foo` is linear
if random() > 0.5 then
    consume(foo);
end if;
consume(foo); -- how?
```

**Reply by hrj** [reply]
More like 
```
let foo: Foo := make(); -- `Foo` is linear
while(random() > 0.1)
  if random() > 0.9 then
    if notIsClosed(foo) then consume(foo);
  else if random() > 0.3 then;
    close(foo);
  end if;
end while
```

**Reply by eudoxia0** [reply]
The short answer is you can't write code with non-deterministic consumption. The rules don't allow it. You have to change your control flow. There's no need to use unsafe code: you can use references to temporarily relax the rules.

The code above can be written more or less like this:

```
let foo: Foo := make(); -- `Foo` is linear
let condition: Bool := True;
let f_or_g: Bool := False
-- `True` means consume using `f`, `False` means consume using `g`.
-- In real life you'd use a union type but I'm trying to keep it simple.
while (random() > 0.1) and condition do
    if random() > 0.5 then
        condition := False;
        f_or_g := True;
    else if random() > 0.3 then
        condition := False;
        f_or_g := False;
    end if;
end while;
if f_or_g then
    f(foo);  -- `foo` is either consumed here...
else
    g(foo); -- ... or here
end if;
```
